### PR TITLE
[SPARK-14837][SQL][STREAMING] Added support in file stream source for reading new files added to subdirs

### PR DESCRIPTION
**Pull request by tdas**
## What changes were proposed in this pull request?

Currently, file stream source can only find new files if they appear in the directory given to the source, but not if they appear in subdirs. This PR add support for providing glob patterns when creating file stream source so that it can find new files in nested directories based on the glob pattern.
## How was this patch tested?

Unit test that tests when new files are discovered with globs and partitioned directories.

### CODE REVIEW

**Comment by tdas** [comment]
@marmbrus @zsxwing

**Comment by marmbrus** [comment]
Why not just support globbing like we do in batch?

**Comment by tdas** [comment]
Globbing? Can you explain that a bit. Code links?

**Comment by marmbrus** [comment]
I.e. something like `/dir/*/*`.  We either do it in DataSource or HDFSFileCatalog.

**Comment by ScrapCodes** [comment]
This is also supported by hadoop. For example, we can pass the globs in sc.textFile("/dir/*/*"). I was wondering, if it will be implemented again.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala [hunk 1]
```diff
@@ -175,25 +175,11 @@ case class DataSource(
...
-              sparkSession,
-              paths = files,
-              userSpecifiedSchema = Some(sourceInfo.schema),
-              className = className,
-              options = new CaseInsensitiveMap(newOptions))
-          Dataset.ofRows(sparkSession, LogicalRelation(newDataSource.resolveRelation()))
-        }
-
         new FileStreamSource(
-          sparkSession, metadataPath, path, sourceInfo.schema, dataFrameBuilder)
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala]: All this code has been moved into the FileStreamSource. It does not make sense that all this file stream specific code should be in the Datasource file.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala [hunk 2]
```diff
@@ -55,10 +62,12 @@ class FileStreamSource(
    */
   private def fetchMaxOffset(): LongOffset = synchronized {
     val filesPresent = fetchAllFiles()
+    logInfo("All files:\n\t" + filesPresent.mkString("\n\t"))
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala]: Remove this line.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala [hunk 3]
```diff
@@ -55,10 +62,12 @@ class FileStreamSource(
...
+
     val newFiles = new ArrayBuffer[String]()
     filesPresent.foreach { file =>
       if (!seenFiles.contains(file)) {
-        logDebug(s"new file: $file")
+        logInfo(s"new file: $file")
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala]: Back to log debug.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala [hunk 4]
```diff
@@ -98,20 +109,49 @@ class FileStreamSource(
...
 
   private def fetchAllFiles(): Seq[String] = {
-    val startTime = System.nanoTime()
-    val files = fs.listStatus(new Path(path))
-      .filterNot(_.getPath.getName.startsWith("_"))
-      .map(_.getPath.toUri.toString)
-    val endTime = System.nanoTime()
-    logDebug(s"Listed ${files.size} in ${(endTime.toDouble - startTime) / 1000000}ms")
-    files
+    logInfo("Path to glob " + pathGlobPattern)
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala]: Remove this.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala [hunk 5]
```diff
@@ -98,20 +109,49 @@ class FileStreamSource(
...
+    val startTime = System.nanoTime
+    val files = SparkHadoopUtil.get.globPath(pathGlobPattern)
+    val endTime = System.nanoTime
+    logInfo(s"Listed ${files.size} in ${(endTime.toDouble - startTime) / 1000000}ms")
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala]: Make this logDebug.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala [hunk 6]
```diff
@@ -98,20 +109,49 @@ class FileStreamSource(
...
+    files.map(_.toUri.toString)
+  }
+
+  private def findBasePath(): Path = {
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala]: Add docs.

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala [hunk 7]
```diff
@@ -439,6 +442,105 @@ class FileStreamSourceSuite extends FileStreamSourceTest with SharedSQLContext {
...
+      testStream(filtered)(
+        // Create new src/level1=xyz/level2=0/ and write to it, should read value for level2 data
+        // but not level1 as it is not in scope of the provided glob pattern src/level1=xyz/*/*
+        AddTextFileData("{'value': 'drop1'}\n{'value': 'keep2'}", level2Dir, tmp),
+        CheckAnswer(("keep2", 0)),
+
+        AddTextFileData("{'value': 'keep3'}", level2Dir, tmp),
+        CheckAnswer(("keep2", 0), ("keep3", 0)),
+
+        // Create new dir/subDir/subsubdir and write to it, should read
```
Comment (tdas) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala]: Update comments.

**Comment by tdas** [comment]
@marmbrus Please take a look once again.

**Comment by tdas** [comment]
@yhuai @marmbrus

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala [hunk 8]
```diff
@@ -97,21 +102,30 @@ class FileStreamSource(
     assert(startId <= endId)
     val files = metadataLog.get(Some(startId + 1), Some(endId)).flatMap(_._2)
     logInfo(s"Processing ${files.length} files from ${startId + 1}:$endId")
-    logDebug(s"Streaming ${files.mkString(", ")}")
-    dataFrameBuilder(files)
+    logTrace(s"Files are:\n\t" + files.mkString("\n\t"))
+    val newOptions = options.filterKeys(_ != "path")
```
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala]: I think our `options` is case preserving. So, when we compare with the `path`, we need to convert the string to its lowercase form.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala [hunk 9]
```diff
@@ -97,21 +102,30 @@ class FileStreamSource(
...
+    val newDataSource =
+      DataSource(
+        sparkSession,
+        paths = files,
+        userSpecifiedSchema = Some(schema),
+        className = fileFormatClassName,
+        options = new CaseInsensitiveMap(newOptions))
```
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala]: do we still need to add `basePath` to this `newOptions`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala [hunk 10]
```diff
@@ -33,12 +35,14 @@ import org.apache.spark.util.collection.OpenHashSet
...
-    metadataPath: String,
     path: String,
+    fileFormatClassName: String,
     override val schema: StructType,
-    dataFrameBuilder: Array[String] => DataFrame) extends Source with Logging {
+    metadataPath: String,
+    options: Map[String, String]) extends Source with Logging {
 
   private val fs = new Path(path).getFileSystem(sparkSession.sessionState.newHadoopConf())
+  private val qualifiedBasePath = fs.makeQualified(new Path(path))
```
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala]: Maybe add a comment to mention that this `qualifiedBasePath` can contain glob patterns?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala [hunk 11]
```diff
@@ -97,21 +102,30 @@ class FileStreamSource(
...
-      .filterNot(_.getPath.getName.startsWith("_"))
-      .map(_.getPath.toUri.toString)
-    val endTime = System.nanoTime()
-    logDebug(s"Listed ${files.size} in ${(endTime.toDouble - startTime) / 1000000}ms")
+    val startTime = System.nanoTime
+    val globbedPaths = SparkHadoopUtil.get.globPathIfNecessary(qualifiedBasePath)
+    val catalog = new ListingFileCatalog(sparkSession, globbedPaths, options, Some(new StructType))
+    val files = catalog.allFiles().map(_.getPath.toUri.toString)
+    val endTime = System.nanoTime
+    logInfo(s"Listed ${files.size} in ${(endTime.toDouble - startTime) / 1000000}ms")
```
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala]: Nice!

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala [hunk 12]
```diff
@@ -444,6 +445,79 @@ class FileStreamSourceSuite extends FileStreamSourceTest with SharedSQLContext {
...
+      val partitionBarSubDir = new File(dir, "partition=bar")
+
+      val schema = new StructType().add("value", StringType).add("partition", StringType)
+      val fileStream = createFileStream("json", s"${dir.getCanonicalPath}/*/*", Some(schema))
+      val filtered = fileStream.filter($"value" contains "keep")
+      val nullStr = null.asInstanceOf[String]
+      testStream(filtered)(
+        // Create new partition=foo sub dir and write to it, should read only value, not partition
+        AddTextFileData("{'value': 'drop1'}\n{'value': 'keep2'}", partitionFooSubDir, tmp),
+        CheckAnswer(("keep2", nullStr)),
```
Comment (yhuai) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala]: why do we have a `nullStr` at here?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala (hunk 8 above)
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala]: Aah right.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala (hunk 9 above)
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala]: No we dont need to. Not adding base path, and giving only a list of files (no dirs in them) ensures that partitioned directories are not parsed. That is the intention, we are not supporting partitioned directories in the file stream yet.

On that note, if the user exlicitly specifies base path in the options, I dont know what will happen. Maybe we should throw an error in that case. What do you think?

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala (hunk 12 above)
Comment (tdas) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala]: I want to make sure in this test that the partition directories are never parsed, even if the user tries to extract it by specifying it in the user defined schema.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala (hunk 10 above)
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala]: okay.

**Comment by yhuai** [comment]
LGTM pending jenkins

**Comment by yhuai** [comment]
Thanks! Merging to master and branch 2.0.

**Pull request closed without merging**